api/auth_middleware_test.go

```diff
@@ -1,63 +1,93 @@
 package api_test
 
 import (
+	"encoding/base64"
 	"fmt"
 	"net/http"
 	"testing"
 
+	"github.com/topfreegames/Will.IAM/models"
 	helpers "github.com/topfreegames/Will.IAM/testing"
 )
 
-func TestAuthMiddlewareKeyPairShouldAuthenticateUser(t *testing.T) {
+func TestAuthMiddleware(t *testing.T) {
 	helpers.CleanupPG(t)
 
-	rootSA := helpers.CreateRootServiceAccountWithKeyPair(t)
-	app := helpers.GetApp(t)
-	req, err := http.NewRequest(http.MethodGet, "/service_accounts", nil)
-
-	if err != nil {
-		t.Errorf("Could not create HTTP request for /service_accounts")
-	}
-
-	req.Header.Set("Authorization", fmt.Sprintf(
-		"keypair %s:%s", rootSA.KeyID, rootSA.KeySecret,
-	))
-
-	response := helpers.DoRequest(t, req, app.GetRouter())
-
-	if response.Code != http.StatusOK {
-		t.Errorf("Expected status %d. Got %d", http.StatusOK, response.Code)
-	}
-
-	serviceAccountName := response.Header().Get("x-service-account-name")
-	if serviceAccountName != "root" {
-		t.Errorf("Expected service account name %s. Got %s", "root", serviceAccountName)
-	}
-}
-
-func TestAuthMiddlewareBearerShouldAuthenticateUser(t *testing.T) {
-	helpers.CleanupPG(t)
-
-	rootSA := helpers.CreateRootServiceAccountWithOAuth(t)
+	oauthSA := helpers.CreateRootServiceAccountWithOAuth(t, "oauthUser", "oauth.user@test.com")
+	keyPairSA := helpers.CreateRootServiceAccountWithKeyPair(t, "keyPairUser", "keypair.user@test.com")
 	tokensRepo := helpers.GetRepo(t).Tokens
-	tokens, _ := tokensRepo.FindByEmail(rootSA.Email)
+	tokens, _ := tokensRepo.FindByEmail(oauthSA.Email)
 	token := tokens[0]
-	app := helpers.GetApp(t)
-	req, err := http.NewRequest(http.MethodGet, "/service_accounts", nil)
 
-	if err != nil {
-		t.Errorf("Could not create HTTP request for /service_accounts")
+	testCases := []struct {
+		name                string
+		serviceAccount      *models.ServiceAccount
+		requestHeaders      map[string]string
+		wantResponseHeaders map[string]string
+		wantResponseCode    int
+	}{
+		{
+			name:                "KeyPairAuthorization",
+			serviceAccount:      keyPairSA,
+			requestHeaders:      map[string]string{"authorization": fmt.Sprintf("keypair %s:%s", keyPairSA.KeyID, keyPairSA.KeySecret)},
+			wantResponseHeaders: map[string]string{"x-service-account-name": "keyPairUser"},
+			wantResponseCode:    http.StatusOK,
+		},
+		{
+			name:                "OAuthAuthorization",
+			serviceAccount:      oauthSA,
+			requestHeaders:      map[string]string{"authorization": fmt.Sprintf("bearer %s", token.AccessToken)},
```

Comment:
nit: prefer "%v" in format strings, unless you a reason not to (eg, "%s" when you want to force something to be formatted as string).

Especially avoid %d and %f if you can - they are easy to break when refactoring.

Same throughout.